MINI-PROJECTS.md

```diff
@@ -104,6 +104,314 @@ share a bound.
+entirely in the front-end (for example, catching `x = x++` as modifying a variable
+twice between sequence points could be done in side effect removal in ansi-c), while
+others may be best to be instrumentation (for example handling the unspecified nature
+of the order of evaluation of function arguments).
+
+Talk to Michael Tautschnig or Martin Brain for this item.
+
+### Reporting Software Errors with CWE Information
+
+The [Common Weakness Enumeration project](http://cwe.mitre.org/) seeks to classify software
+weaknesses and errors in a common dictionary. Reporting CWE information together with failed
+assertions, such as [Free of Pointer not at Start of Buffer](http://cwe.mitre.org/data/definitions/761.html),
+would help both users of the tools as well as industrial partners. In the first instance, errors
+already caught by present instrumentation should be augmented with additional information. In a
+second step, instrumentation for other problems listed in the CWE database should be developed.
+
+### Machine Learn Refinement settings
+
+The refinement option enables a tactic of under and over approximation to try to solve the
+generated formulae more quickly. There are a large number of possible options – which operations
+to approximate, how much to under or over approximate them, how much to add back if this doesn't
+work, etc. Rather than trying to manually pick the best strategy from these, it would be good
+to machine learn them.
+
+It would probably be worth asking Nando de Freitas' advice on the machine learning side of things.
+
+Ruben is now looking at this one. (?)
+
+### Cluster CBMC
+
+Most users of CBMC have access to multi-core machines; many have clusters available. There
+are a number of different approaches and a number of different levels at which the task
+could be parallelised; per assertion, portfolio approaches of different back-ends,
+parallel solvers, etc. It would be good to be able to exploit these. The caching approach
+suggested under “KLEE-PROVER” might be an easy route to implementing this.
+
+### x86 to Goto
+
+Look at various efforts to model x86 assembler, e.g., Andrew Kennedy,
+“Formalizing .EXEs, .DLLs, and all that”, Gang Tan, “Reusable tools for formal modeling of machine code”,
+and check whether this can turn into a translator from x86 to goto binary.
+
+Also look at Google's Native Client, and talk to Matt, who has tried Valgrind.
+
+Furthermore, talk to Michael, who has got a student who made some progress on this.
+
+And maybe have a read of Automated synthesis of symbolic instruction encodings from I/O samples.
+
+### Branch Prediction
+
+Implement branch prediction algorithms in path-symeex.
+
+### An abstract interpreter for C
+
+There's a lack for a robust, open source, abstract interpreter for C/C++. CPROVER contains most of the
+components for it, the key thing that would be needed is a front-end and some testing, plus possibly
+some more domains.
+
```

Comment:
This is `goto-analyzer`.  We can remove this project suggestion -- turns out to not be so mini after all!